File: archived/ptune-log/src/features/google_tasks/services/export/MarkdownTaskParser.ts
import { MarkdownTaskEntry } from 'src/core/models/tasks/MarkdownTaskEntry';
import { logger } from 'src/core/services/logger/loggerInstance';

/**
 * Markdown 形式のタスク一覧を解析し、ParsedTask オブジェクト配列に変換するクラス
 */
export class MarkdownTaskParser {
  /**
   * タスク行を解析して ParsedTask[] を返す（親子関係をインデントで判定）
   */
  static parse(lines: string[]): MarkdownTaskEntry[] {
    const taskRe = /^\s*-\s\[\s\]\s+(.*?)(?:\s*🍅x?(\d+))?$/;
    const result: MarkdownTaskEntry[] = [];
    let currentParentIndex: number | null = null;

    for (const [i, line] of lines.entries()) {
      const match = line.match(taskRe);
      if (!match) continue;

      const title = match[1].trim();
      if (!title) continue;

      const pomodoro = match[2] ? parseInt(match[2]) : 0;
      const indent = line.search(/\S|$/);

      const task: MarkdownTaskEntry = {
        index: i,
        title,
        pomodoro,
        rawLine: line,
      };

      if (indent > 0 && currentParentIndex !== null) {
        task.parent_index = currentParentIndex;
      } else {
        currentParentIndex = i;
      }

      result.push(task);
    }

    logger.debug(`[MarkdownTaskParser.parse] parsed=${result.length}`);
    return result;
  }
}

File: archived/PtuneSync/PtuneSync/GoogleTasks/MarkdownTaskParser.cs
using System.Text.RegularExpressions;
using PtuneSync.Models;
using PtuneSync.Infrastructure;
using System.Collections.Generic;
using System.Linq;

namespace PtuneSync.GoogleTasks;

public static class MarkdownTaskParser
{
    private static readonly Regex TaskRe = new(@"^\s*-\s\[\s\]\s+(.*?)(?:\s*🍅x?(\d+))?$");

    public static List<ParsedTask> Parse(IEnumerable<string> lines)
    {
        var result = new List<ParsedTask>();
        int? currentParent = null;
        int index = 0;

        foreach (var line in lines)
        {
            var m = TaskRe.Match(line);
            if (!m.Success) continue;

            var title = m.Groups[1].Value.Trim();
            if (string.IsNullOrEmpty(title)) continue;

            int.TryParse(m.Groups[2].Value, out int pomo);
            int indent = line.TakeWhile(char.IsWhiteSpace).Count();

            var task = new ParsedTask
            {
                Index = index++,
                Title = title,
                Pomodoro = pomo,
                RawLine = line
            };

            if (indent > 0 && currentParent.HasValue)
                task.ParentIndex = currentParent.Value;
            else
                currentParent = task.Index;

            result.Add(task);
        }

        AppLog.Debug("[MarkdownTaskParser] Parsed {0} tasks", result.Count);
        return result.AsEnumerable().Reverse().ToList();
    }
}
